stackslib/src/chainstate/nakamoto/miner.rs

```diff
@@ -653,70 +659,98 @@ impl BlockBuilder for NakamotoBlockBuilder {
-                                        &total_budget
-                                    );
-                                    return TransactionResult::skipped_due_to_error(
-                                        &tx,
-                                        Error::BlockTooBigError,
-                                    );
-                                }
-                            }
-                            _ => return TransactionResult::error(&tx, e),
+
+            let (fee, receipt) =
+                match StacksChainState::process_transaction(clarity_tx, tx, quiet, ast_rules) {
+                    Ok(x) => x,
+                    Err(e) => {
+                        return parse_process_transaction_error(clarity_tx, tx, e);
+                    }
+                };
+
+            let mut soft_limit_reached = false;
+            // We only attempt to apply the soft limit to non-boot code contract calls.
+            if non_boot_code_contract_call {
+                if let Some(percentage) = self.cost_limit_percentage {
+                    if percentage < 100 {
+                        let remaining_limit = clarity_tx
+                            .block_limit()
+                            .expect("Failed to obtain block limit from miner's block connection");
+                        let mut block_limit = remaining_limit.clone();
+                        // Attempt to spread the cost limit across the tenure by always using only 25% of the remaining budget
+                        // until the remaining budget is less than 100 in which case we just use 100% of the remaining budget
+                        if block_limit.divide(100).is_ok() {
+                            block_limit.multiply(25).expect(
+                                "BUG: Successfully divided by 100, but failed to multiply block limit by 25",
+                            );
+                            clarity_tx.reset_cost(block_limit);
+                            soft_limit_reached = matches!(
+                                StacksChainState::process_transaction(
+                                    clarity_tx, tx, quiet, ast_rules
+                                ),
+                                Err(Error::CostOverflowError(_, _, _))
+                            );
+                            clarity_tx.reset_cost(remaining_limit);
                         }
                     }
                 }
-            };
+            }
+
             info!("Include tx";
                   "tx" => %tx.txid(),
                   "payload" => tx.payload.name(),
                   "origin" => %tx.origin_address());
 
             // save
             self.txs.push(tx.clone());
-            TransactionResult::success(&tx, fee, receipt)
+            TransactionResult::success(&tx, fee, receipt, soft_limit_reached)
         };
 
         self.bytes_so_far += tx_len;
         result
     }
 }
+
+fn parse_process_transaction_error(
```

Comment:
Maybe I should still apply this in the soft limit case so we can take advantage of the proportion heuristic and do something similar to the marking of transactions/block limits in the outer loop.

Reply:
Granted this already is applied to transactions initially and only if they pass this check do we do the soft limit check so prob is fine...